Fairness metrics

superp0tat0 opened the pull request:
Create the metric draft
(Only support groupwise absolute difference right now)

mllg commented:
I've fixed some roxygen warnings in #6d09513. The code looks nearly complete so this PR could be merged soon, but it still lacks some documentation. I'll add comments here in this PR. 

Additionally, having an example or (maybe even better) a short vignette would be nice to showcase the measures. But this could also go into a different PR.

superp0tat0 commented:
> I've fixed some roxygen warnings in #6d09513. The code looks nearly complete so this PR could be merged soon, but it still lacks some documentation. I'll add comments here in this PR.
> 
> Additionally, having an example or (maybe even better) a short vignette would be nice to showcase the measures. But this could also go into a different PR.

I have added some examples at the [wiki page](https://github.com/mlr-org/mlr3fairness/wiki/Fairness-Metric-Demo). I could make another short vignette later also.

mllg commented:
I also just noticed that you are not populating the `mlr_measures` dictionary but have a custom solution in place. 

Here is an example on how to register the measures during package load:

https://github.com/mlr-org/mlr3proba/blob/main/R/zzz.R#L86 f

mllg commented:
> I have added some examples at the [wiki page](https://github.com/mlr-org/mlr3fairness/wiki/Fairness-Metric-Demo). I could make another short vignette later also.

Ah, haven't seen this. I would still recommend to turn this in a proper vignette to let R CMD check run the code automatically and ensure that the example still works with the latest version.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+#' the equalized odds ratios. More examples could viewed on mlr3fairness wikipage.
+#'
+#' Predefined measures can be found in the [dictionary][mlr3misc::Dictionary] [mlr_measures].
+#' Predefined operations can be found in the [mlr3fairness].
```
some incomplete sentences here.

superp0tat0 replied:
Not sure what "incomplete" means. Do you want me to add some clarifications on measures and operations?

pfistfl replied:
`for Fairness Measure problems` -> t`to allow for measuring fairness`.

superp0tat0 replied:
Got it, corrected my grammar.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+MeasureFairness = R6Class("MeasureFairness", inherit = Measure, cloneable = FALSE,
+  public = list(
+    #' @template field_fun
+    fun = NULL,
```
Not clear what this is supposed to be.

superp0tat0 replied:
Those are the templates stored at man-roxygen. I followed the documentation styles in mlr3.

superp0tat0 replied:
Should I remove them?

pfistfl replied:
I think If you do not need/use the slot just get rid of it yes.

mllg replied:
Sorry, my comment was not clear. I was wondering what the public field `fun` is supposed to store. The documentation just states `"This function referred to the fairness measure operation functions."`. I don't think that this is helpful for the user. Maybe this field should be private?

superp0tat0 replied:
I see what you mean. I guess I will change this field to private.

superp0tat0 replied:
Also this field is needed to store the operation function, so we cannot get rid of it.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #' @param base_measure The measure used to perform fairness operations.
```
Stating the argument type ([Measure] in this case) would help. We try to always do this, c.f. mlr3 man pages.

superp0tat0 replied:
Corrected

mllg replied:
> Corrected

Did you forget to push? Where did you correct it?

superp0tat0 replied:
I haven't push it yet, should be up here now.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
```
What are possible values?

pfistfl replied:
Basically state the three options (I guess they are supplied as a character?)

superp0tat0 replied:
Fixed

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
+    #' @param ps The parameter sets.
```
`ps` is not used and can be removed.

superp0tat0 replied:
Fixed

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+        man = paste0("mlr_measures_fairness")
+      )
+      self$fun = get(operation, envir = asNamespace("mlr3fairness"), mode = "function")
+      self$base_measure = base_measure
```
Since this is user input, it makes sense to validate it and give a proper error message if this is not a measure.
```r
self$base_measure =  `assert_measure(base_measure)`
```

mllg replied:
Same for the argument `operation`, otherwise the error message will be hard to understand

```r
assert_choice(operation, c("groupwise_abs_diff", "groupwise_diff", "groupwise_quotient"))
```

superp0tat0 replied:
Fixed, thanks for the comments.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
+    #' @param ps The parameter sets.
+    initialize = function(operation, base_measure, ps = ps()) {
```
`operation` should maybe have a default value?

superp0tat0 replied:
set the default operation to "groupwise_abs_diff"

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+
+  private = list(
+    .score = function(prediction, task, ...) {
+      assert_prediction(prediction)
```
this assertion is done in the parent class.

superp0tat0 replied:
Fixed

superp0tat0 replied:
Got one question, the parent class (`Measure()`) do have the assert() for "tasks" but MeasureFairness did not inherit those assert. Instead, it directly passed the arguments to invoke functions.

superp0tat0 replied:
Found the problem, mark as resolved

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+  private = list(
+    .score = function(prediction, task, ...) {
+      assert_prediction(prediction)
+      if ("requires_task" %in% self$properties && is.null(task)) {
```
this is also done in the parent class.

superp0tat0 replied:
Fixed, delete the assertions.

superp0tat0 replied:
Same question as above

superp0tat0 replied:
I found where the problem is, mark as resolved.

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+        stopf("Measure '%s' requires a task", self$id)
+      }
+
+      invoke(self$fun, prediction = prediction, na_value = self$na_value, data_task = task,
```
There is no `self$na_value`.

superp0tat0 replied:
Fixed

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+  )
+)
+
+mlr_measures$add("mlr_measures_fairness", MeasureFairness, name = "mlr_measures_fairness")
```
maybe just call it `"fairness"`?

superp0tat0 replied:
Fixed

mllg commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+      if ("requires_task" %in% self$properties && is.null(task)) {
+        stopf("Measure '%s' requires a task", self$id)
+      }
+
```
we should enforce some assertion here: 

* task has a single designated column with role `pta`
* there are not more than 2 levels in the `pta` column (AFAICT this measure will not work for more groups, right?)

pfistfl replied:
Let us not do the 2nd assertion where we compute the difference. 
Then a `max_abs_diff` measure could e.g. work for >2 groups.

mllg replied:
Are you arguing to add `max_abs_diff` as operation or to change all operations to operate on more than 2 groups? We can move the assertion to a different place, but it should be performed at some point for operations which do not support multiple groups.

superp0tat0 replied:
I think we could put the assertion into the operation functions. For example, groupwise_abs_diff(). However, should we change the function name to reflect its supported pta? Like binary_abs_diff etc.

superp0tat0 replied:
I have added an assertions.R file to R holder. Which contains assert_binary_pta and assert_pta. assert_binary_pta will call assert_pta everytime and all the fairness operation will call assert_binary_pta to check the binary pta field.

mllg commented on R/groupwise_abs_diff.R:
```diff
@@ -0,0 +1,24 @@
+  prediction = split(as.data.table(prediction), data_task$data(cols = subcol))
+  prediction = map(prediction, as_prediction_classif)
+  msr1 = prediction[[1]]$score(base_measure)
+  msr2 = prediction[[2]]$score(base_measure)
```
the 5 above lines are repeated in all 3  `groupwise_*` functions. this should probably go into a helper function then.

superp0tat0 replied:
Fixed, create helper function binary_measure_score in helper.R

superp0tat0 replied:
Fixed, create helper function binary_measure_score in helper.R.
However, we may need more helper functions in case of non binary case. But that would some further problems.

mllg commented on R/groupwise_abs_diff.R:
```diff
@@ -0,0 +1,24 @@
+}
+
+#' @include measures.R
+add_measure(groupwise_abs_diff, "Groupwise Absolute Difference", "regr", 0, Inf, FALSE)
```
this does not make sense to me, maybe I'm missing something? `groupwise_abs_diff` is not a measure itself, but a method to aggregate multiple scores. I don't see how this can have a task type, lower+upper bound etc.

superp0tat0 replied:
Fixed, I think I misunderstand something before.

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+#' Predefined measures can be found in the [dictionary][mlr3misc::Dictionary] [mlr_measures].
+#' Predefined operations can be found in the [mlr3fairness].
+#'
+#' @export
```
We are definitely missing an example here. 
Consider adding an example of how to construct 1-2 common measures using this approach.

superp0tat0 replied:
Added example about Predictive Parity.

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+#' This measure specializes [Measure] for Fairness Measure problems:
+#' Users could use [groupwise_abs_diff()], [groupwise_diff()], or [groupwise_quotient()] to evaluate
+#' the fairness measures based on base_measures. For example, the false positive rate bias or
+#' the equalized odds ratios. More examples could viewed on mlr3fairness wikipage.
```
Do not reference the wiki in the docs, instead either point to a vignette or to the examples section.

superp0tat0 replied:
Are there any vignette examples I could review?

pfistfl replied:
https://github.com/mlr-org/mcboost/tree/main/vignettes

Most of the `mlr` vignettes have been integrated in the mlr3 book, that is why mlr often does not have any.
We will at a later point perhaps also add the fairness vignettes to mrl3measures, but for now let us develop them with the package.

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+#' @description
+#' This measure specializes [Measure] for Fairness Measure problems:
+#' Users could use [groupwise_abs_diff()], [groupwise_diff()], or [groupwise_quotient()] to evaluate
+#' the fairness measures based on base_measures. For example, the false positive rate bias or
```
You are using base_measures as a word but it is not clear what this means.
Trying to improve the description:

```
This measure specializes [Measure] to ...
A common approach to quantifying a model's fairness is to compute the difference between a protected and an unprotected group according to some performance metric such as `classification error`, `false positive rate` or others. 
This measure allows for the composition of a fairness-metric from such a `base_measure` by computing group-wise difference according to some strategy. 
Several options wrt. computing difference such as ... are available.

pfistfl replied:
I understand that this is really hard, but if you try to put yourself in the shoes of somebody who has not written the code, it needs to be understandable.

pfistfl replied:
A good sentence to also include here is that this implements `statistical group fairness measures`

superp0tat0 replied:
That is really helpful, I have updated the documentation

superp0tat0 replied:
Mark as resolved

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
```
We always annotate the type see e.g. here: https://github.com/mlr-org/mlr3proba/blob/7f7dc61a29d7fd91849027eb971e7e0765cfb0c7/R/pecs.R#L13

superp0tat0 replied:
Fixed

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
+    #' @param ps The parameter sets.
+    initialize = function(operation, base_measure, ps = ps()) {
```
Add and document a default for `operation` here, I guess `"abs_diff"`.

pfistfl replied:
If we are not using the `ps` the remove it.

superp0tat0 replied:
Fixed

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    #' @param base_measure The measure used to perform fairness operations.
+    #' @param operation The operation name performed on the base measures.
```
Document the default.

superp0tat0 replied:
Fixed

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+      super$initialize(
+        id = paste0("fairness.", base_measure$id),
+        range = c(-Inf, Inf),
+        minimize = FALSE,
```
Ok, so here things become difficult `range` depends on `operation`. But `minimize` should always be `TRUE` I think?

mllg replied:
I'm wondering if it would be easier to have three separate measures then, one for each operation?

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,54 @@
+        minimize = FALSE,
+        predict_type = base_measure$predict_type,
+        packages = "mlr3fairness",
+        man = paste0("mlr_measures_fairness")
```
You do not need the paste0

superp0tat0 replied:
Fixed

pfistfl commented on R/groupwise_abs_diff.R:
```diff
@@ -0,0 +1,24 @@
+#' @param ... Further arguments, currently ignored.
+#'
+#' @return Absolute difference of base measues between binary protected groups.
+#' @export
```
Does not need to be exported.

superp0tat0 replied:
Fixed

pfistfl commented on R/groupwise_abs_diff.R:
```diff
@@ -0,0 +1,24 @@
+#' @param data_task The data task for the fairness metric.
+#' @param ... Further arguments, currently ignored.
+#'
+#' @return Absolute difference of base measues between binary protected groups.
```
We always document the data type of params and outputs, see above.

superp0tat0 replied:
Fixed

pfistfl commented on R/groupwise_diff.R:
```diff
@@ -0,0 +1,23 @@
+#' @param ... Further arguments, currently ignored.
+#'
+#' @return Difference of base measues between binary protected groups.
+#' @export
```
No export, data types. But for non-exported functions just drop change `#'` to `#`. This holds for all functions that are internally used in `MeasureFairness`.

superp0tat0 replied:
Fixed

pfistfl commented on R/measures.R:
```diff
@@ -0,0 +1,27 @@
+measures = new.env(parent = emptyenv())
+
+# adds items to registry
+add_measure = function(obj, title, type, lower, upper, minimize) {
```
So the measures registry is `mlr_measures` and this is available if you import `mlr_measures`.
You can simply add to this registry using `mlr_measures$add`. 
See https://github.com/mlr-org/mlr3proba/blob/7f7dc61a29d7fd91849027eb971e7e0765cfb0c7/R/zzz.R#L123 for an example on how proba adds to mlr_pipeops

pfistfl replied:
This means, that this whole file is most likely not needed.

superp0tat0 replied:
Deleted the profile, sorry i just forgot to delete it.

superp0tat0 commented on R/MeasureFairness.R:
```diff
@@ -44,6 +45,7 @@ MeasureFairness = R6Class("MeasureFairness", inherit = Measure, cloneable = FALS
       super$initialize(
         id = paste0("fairness.", base_measure$id),
         range = c(-Inf, Inf),
+        properties = "reqiures_task",
```
This properties do not pass in the measures somehow. eg:
```r
me = MeasureFairness$new("groupwise_abs_diff", base_measure)
me$properties
>>> character(0)
```

@mllg Could you please take a look? It took me about an hour to debug but I cannot find the reasons.

mllg replied:
This works for me, maybe you forgot to reload the package?

pfistfl replied:
@superp0tat0 check this works in the unit tests

superp0tat0 replied:
I see. Resolved the issue after I rebuild and install the whole package.

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,66 @@
+#' @name mlr_measures_fairness
+#'
+#' @description
+#' This measure specializes [Measure] to allow for measuring statistical group fairness:
```
Please check that the link to measure works. I would assume it needs an `(mlr3)`?

superp0tat0 replied:
Fixed the hyperlink

pfistfl commented:
We are currently missing unit tests for the different functions. Please try to include those aswell.

pfistfl commented:
Open Todos: 
- Write a vignette on how to measure fairness `usethis::use_vignette()` can give you a template
- Write detailed unit tests for all functions, measure properties, tasks etc.
- I would suggest that `mlr3fairness` comes with a fixed set of widely used metrics (the top 10 metrics we identified for now). 
  those have to be added to `zzz.R` and documented.

pfistfl commented:
I have added a few more pointers to get started

superp0tat0 commented:
> Open Todos:
> 
> * Write a vignette on how to measure fairness `usethis::use_vignette()` can give you a template
> * Write detailed unit tests for all functions, measure properties, tasks etc.
> * I would suggest that `mlr3fairness` comes with a fixed set of widely used metrics (the top 10 metrics we identified for now).
>   those have to be added to `zzz.R` and documented.

No problem, working on it.

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,85 @@
+#' @examples
+#' # Predfined measures:
+#' msr("fairness.fpr")
+NULL
```
missing a newline here

pfistfl commented on R/MeasureFairness.R:
```diff
@@ -0,0 +1,85 @@
+#'
+#' A set of widely used metrics is included in mlr3fairness.
+#' * `fairness.fpr` :: Abs. difference in false positive rates across groups
+#' * `fairness.acc` :: Abs. difference in accuracy across groups
```
add and implement more here

mllg commented on tests/testthat/test-test_datasets.R:
```diff
@@ -1,36 +1,53 @@
-
+check_data_format = function(data) {
+  expect_true(is.data.frame(data))
+  expect_true(sum(colnames(data) == tolower(colnames(data))) == length(colnames(data))) #Check all True
```
use `all()`

superp0tat0 replied:
Got it, I was looking for this function but haven't find it.

mllg commented on R/operations.R:
```diff
@@ -0,0 +1,57 @@
+#' @param ... Further arguments, currently ignored.
+#'
+#' @return Absolute difference of base measues between binary protected groups.
+groupwise_abs_diff <- function(prediction, base_measure, data_task, ...) {
```
you could re-use the other function here:

```r
groupwise_abs_diff <- function(prediction, base_measure, data_task, ...) {
  abs(groupwuse_diff(prediction, base_measure, data_task, ...))
}
```

mllg commented on tests/testthat/test-test_datasets.R:
```diff
@@ -1,36 +1,53 @@
+  assert_col_type = (sapply(compas_data, class) == c("factor", "integer", "factor", "factor",
+                                                    "integer", "integer", "factor", "integer",
+                                                    "integer", "factor", "factor", "factor"))
+  expect_true(sum(assert_col_type) == 12L)
```
`all()`

superp0tat0 replied:
Solved

mllg reviewed (approved):
minor style suggestions
